esp-hal/src/rtc_cntl/mod.rs

```diff
@@ -257,13 +273,114 @@ impl<'d> Rtc<'d> {
+        //   esp32s2, esp32s3
+
+        // - LP_AON is used on the following chips: esp32c5, esp32c6, esp32c61, esp32h2
+
+        // For registers and peripherals used in esp-idf, see https://github.com/search?q=repo%3Aespressif%2Fesp-idf+RTC_BOOT_TIME_LOW_REG+RTC_BOOT_TIME_HIGH_REG+path%3A**%2Frtc.h&type=code
+
+        #[cfg(not(any(esp32c6, esp32h2)))]
+        let rtc_cntl = unsafe { &*LPWR::ptr() };
+        #[cfg(any(esp32c6, esp32h2))]
+        let rtc_cntl = unsafe { &*LP_AON::ptr() };
+
+        let (l, h) = (rtc_cntl.store2(), rtc_cntl.store3());
+
+        let l = l.read().bits() as u64;
+        let h = h.read().bits() as u64;
+
+        // https://github.com/espressif/esp-idf/blob/23e4823f17a8349b5e03536ff7653e3e584c9351/components/newlib/port/esp_time_impl.c#L115
+        l + (h << 32)
     }
 
     /// Read the current value of the rtc time registers in milliseconds.
+    fn get_boot_time_ms(&self) -> u64 {
+        self.get_boot_time_us() / 1_000
+    }
+
+    /// Set the current value of the boot time registers in microseconds.
+    fn set_boot_time_us(&self, boot_time_us: u64) {
+        // Please see `get_boot_time_ms` for documentation on registers and peripherals
+        // used for certain SOCs.
+
+        #[cfg(not(any(esp32c6, esp32h2)))]
+        let rtc_cntl = unsafe { &*LPWR::ptr() };
+        #[cfg(any(esp32c6, esp32h2))]
+        let rtc_cntl = unsafe { &*LP_AON::ptr() };
+
+        let (l, h) = (rtc_cntl.store2(), rtc_cntl.store3());
+
+        // https://github.com/espressif/esp-idf/blob/23e4823f17a8349b5e03536ff7653e3e584c9351/components/newlib/port/esp_time_impl.c#L102-L103
+        l.write(|w| unsafe { w.bits((boot_time_us & 0xffffffff) as u32) });
+        h.write(|w| unsafe { w.bits((boot_time_us >> 32) as u32) });
+    }
+
+    /// Read the current value of the time registers in microseconds.
+    pub fn get_time_us(&self) -> u64 {
+        // current time is boot time + time since boot
+        self.get_boot_time_us() + self.get_rtc_time_us()
+    }
+
+    /// Read the current value of the time registers in milliseconds.
     pub fn get_time_ms(&self) -> u64 {
-        self.get_time_raw() * 1_000 / RtcClock::get_slow_freq().frequency().to_Hz() as u64
+        // current time is boot time + time since boot
+        self.get_boot_time_ms() + self.get_rtc_time_ms()
+    }
+
+    /// Set the current value of the time registers in microseconds.
+    ///
+    /// This function will fail if `time_us` is less than the time
+    /// since boot. This happens because when setting the time, we
+    /// have to set the boot time which is calculated by subtracting
+    /// `time_us` from the time since boot. However, since time is
+    /// stored with unsigned integers, if `time_us` is less than
+    /// the time since boot, the subtraction result will be negative
+    /// and thus an overflow will occur.
+    pub fn set_time_us(&self, time_us: u64) -> Result<(), RtcSetOverflow> {
```

Comment:
Just a thought but instead of overflowing could this just wrap around instead?

I say this because the implementation of `get_time` (`self.get_boot_time_ms() + self.get_rtc_time_ms()`) can overflow as well but that doesn't return a `Result`.

Reply:
Good point. I'll implement wrapping for both get_time and set_time